lib/web/command_utils.go

```diff
@@ -0,0 +1,100 @@
+/*
+
+ Copyright 2023 Gravitational, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+
+*/
+
+package web
+
+import (
+	"encoding/json"
+	"io"
+	"net"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/trace"
+)
+
+// WSConn is a gorilla/websocket minimal interface used by our web implementation.
```

Comment:
Are all of these necessary? I worry that this basically pins our API to match that of gorilla/websocket, which is not actively maintained and will need to be replaced at some point.

Reply:
The goal here is to replace `Close()` implementation. I'm not aware of any mechanism in Go that allows you to do that without creating an interface. I'm open to alternatives solution if you have one.

Reply:
Can you embed the conn and only override close? Something like

```
type UnclosableWebsocket {
    websocket.Conn
}

func (UnclosableWebsocket) Close() error { return nil }
```

Reply:
This is what I did https://github.com/gravitational/teleport/blob/b46771bf70401af6d16fa7e7af74ac2faf50a7b5/lib/web/command_utils.go#L92-L99
but I still need this interface for function definition, so `CommandHandler` can use the no-close version, and `TerminalHandler` can use the `websocket.Conn`.